src/ControlRecommender.hpp

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2015 - 2022, Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef CONTROLRECOMMENDER_HPP_INCLUDE
+#define CONTROLRECOMMENDER_HPP_INCLUDE
+
+#include <memory>
+#include <vector>
+#include <string>
+
+struct geopm_request_s; // Defined in "geopm/PlatformIO.hpp"
+
+namespace geopm
+{
+    /// @brief Abstract class that recommends control settings
+    class ControlRecommender
+    {
+        public:
+            ControlRecommender() = default;
+            virtual ~ControlRecommender() = default;
+            /// @brief Factory constructor
+            /// @param [in] algorithm Name of algorithm to select
+            /// @returns Unique pointer to a ControlRecommender objecct
+            static std::unique_ptr<ControlRecommender> make_unique(const std::string &algorithm);
+            /// @brief Set the performance bias
+            /// @param [in] bias Abstract number between 0 and 1: a
+            ///         value of 1 is most biased toward performance,
+            ///         and 0 is most biased towards energy efficiency.
+            virtual void performance_bias(double bias) = 0;
```

Comment:
This API is insufficient to enable an agent to express information that is gathered from the TreeComm.  There is no way to have the ControlRecommender send data to other nodes via the TreeComm.  This makes it difficult for distributed agent algorithms to fit into this API.

Reply:
Some additional context from that discussion:

In the meeting we discussed one solution may be for control recommenders to describe things that they'd send over TreeComm. We noted that TreeComm needs to know information about what it will communicate early in launch (i.e., how much space to allocate for samples and policies). If we go that direction, it might be worth evaluating whether there are any communication overhead aspects to consider (e.g., whether we also need to independently enable/disable communicators associated with different recommenders)